stellargraph/layer/knowledge_graph.py

```diff
@@ -191,3 +191,126 @@ def build(self):
         x_out = self(x_inp)
 
         return x_inp, x_out
+
+
+class DistMultScore(Layer):
+    """
+    DistMult scoring Keras layer.
+
+    Original Paper: Embedding Entities and Relations for Learning and Inference in Knowledge
+    Bases. Bishan Yang, Wen-tau Yih, Xiaodong He, Jianfeng Gao, Li Deng. ICLR 2015
+
+    This combines subject, relation and object embeddings into a score of the likelihood of the
+    link.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def build(self, input_shape):
+        self.built = True
+
+    def call(self, inputs):
+        e1, r, e2 = inputs
+        # y_(e_1)^T M_r y_(e_2), where M_r = diag(w_r) is a diagonal matrix
+        score = tf.reduce_sum(e1 * r * e2, axis=2)
+        return score
+
+
+@experimental(reason="results from the reference paper have not been reproduced yet")
+class DistMult:
+    """
+    Embedding layers and a DistMult scoring layers that implement the DistMult knowledge graph
+    embedding algorithm as in https://arxiv.org/pdf/1412.6575.pdf
+
+    Args:
+        generator (KGTripleGenerator): A generator of triples to feed into the model.
+
+        k (int): the dimension of the embedding (that is, a vector in R^k is learnt for each node
+            and each link type)
+
+        embedding_initializer (str or func, optional): The initialiser to use for the embeddings.
+
+        embedding_regularizer (str or func, optional): The regularizer to use for the embeddings.
+    """
+
+    def __init__(
+        self, generator, k, embedding_initializer=None, embedding_regularizer=None,
```

Comment:
the paper mentions constraining the embeddings to have unit length norm. To keep the defaults consistent with the paper you could add in a default constraint here.

```suggestion
        self, generator, k, embedding_initializer=None, embedding_regularizer=None, embedding_constraint=constraints.UnitNorm(axis=0),
```

Reply:
Ah, yeah, nice catch. 👍

Reply:
Ah, unfortunately this doesn't work due to https://github.com/tensorflow/tensorflow/issues/33755: it hits `RuntimeError: Cannot use a constraint function on a sparse variable.`.

<details><summary>Full trace</summary>

```
_____________________________________________________________________________________________________________________________________________________________________________ test_dismult _____________________________________________________________________________________________________________________________________________________________________________

knowledge_graph = <stellargraph.core.graph.StellarDiGraph object at 0x107c67cc0>

    def test_dismult(knowledge_graph):
        # this test creates a random untrained model and predicts every possible edge in the graph, and
        # compares that to a direct implementation of the scoring method in the paper
        gen = KGTripleGenerator(knowledge_graph, 3)
    
        # use a random initializer with a large range, so that any differences are obvious
        init = initializers.RandomUniform(-1, 1)
        x_inp, x_out = DistMult(gen, 5, embeddings_initializer=init).build()
    
        model = Model(x_inp, x_out)
    
        model.compile(loss=losses.BinaryCrossentropy(from_logits=True))
    
        every_edge = itertools.product(
            knowledge_graph.nodes(),
            knowledge_graph._edges.types.pandas_index,
            knowledge_graph.nodes(),
        )
        df = triple_df(*every_edge)
    
        # check the model can be trained on a few (uneven) batches
        model.fit(
            gen.flow(df.iloc[:7], negative_samples=2),
>           validation_data=gen.flow(df.iloc[7:14], negative_samples=3),
        )

tests/layer/test_knowledge_graph.py:113: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training.py:728: in fit
    use_multiprocessing=use_multiprocessing)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py:324: in fit
    total_epochs=epochs)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py:123: in run_one_epoch
    batch_outs = execution_function(iterator)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2_utils.py:86: in execution_function
    distributed_function(input_fn))
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/eager/def_function.py:457: in __call__
    result = self._call(*args, **kwds)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/eager/def_function.py:503: in _call
    self._initialize(args, kwds, add_initializers_to=initializer_map)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/eager/def_function.py:408: in _initialize
    *args, **kwds))
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py:1848: in _get_concrete_function_internal_garbage_collected
    graph_function, _, _ = self._maybe_define_function(args, kwargs)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py:2150: in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py:2041: in _create_graph_function
    capture_by_value=self._capture_by_value),
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/framework/func_graph.py:915: in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/eager/def_function.py:358: in wrapped_fn
    return weak_wrapped_fn().__wrapped__(*args, **kwds)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2_utils.py:73: in distributed_function
    per_replica_function, args=(model, x, y, sample_weights))
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/distribute/distribute_lib.py:760: in experimental_run_v2
    return self._extended.call_for_each_replica(fn, args=args, kwargs=kwargs)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/distribute/distribute_lib.py:1787: in call_for_each_replica
    return self._call_for_each_replica(fn, args, kwargs)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/distribute/distribute_lib.py:2132: in _call_for_each_replica
    return fn(*args, **kwargs)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/autograph/impl/api.py:292: in wrapper
    return func(*args, **kwargs)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2_utils.py:264: in train_on_batch
    output_loss_metrics=model._output_loss_metrics)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_eager.py:311: in train_on_batch
    output_loss_metrics=output_loss_metrics))
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_eager.py:272: in _process_single_batch
    model.optimizer.apply_gradients(zip(grads, trainable_weights))
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/keras/optimizer_v2/optimizer_v2.py:441: in apply_gradients
    kwargs={"name": name})
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/distribute/distribute_lib.py:1917: in merge_call
    return self._merge_call(merge_fn, args, kwargs)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/distribute/distribute_lib.py:1924: in _merge_call
    return merge_fn(self._strategy, *args, **kwargs)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/keras/optimizer_v2/optimizer_v2.py:485: in _distributed_apply
    var, apply_grad_to_update_var, args=(grad,), group=False))
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/distribute/distribute_lib.py:1530: in update
    return self._update(var, fn, args, kwargs, group)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/distribute/distribute_lib.py:2142: in _update
    return self._update_non_slot(var, fn, (var,) + tuple(args), kwargs, group)
../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/distribute/distribute_lib.py:2148: in _update_non_slot
    result = fn(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

var = <tf.Variable 'DISTMULT_NODE/embeddings:0' shape=(4, 5) dtype=float32, numpy=
array([[ 0.00609207,  0.94447637,  0.8121...223864 , -0.5335603 ],
       [ 0.00211883, -0.8434782 , -0.61788964,  0.14921594,  0.03648853]],
      dtype=float32)>, grad = <tensorflow.python.framework.indexed_slices.IndexedSlices object at 0x14afaa208>

    def apply_grad_to_update_var(var, grad):
      """Apply gradient to variable."""
      if isinstance(var, ops.Tensor):
        raise NotImplementedError("Trying to update a Tensor ", var)
    
      apply_kwargs = {}
      if isinstance(grad, ops.IndexedSlices):
        if var.constraint is not None:
          raise RuntimeError(
>             "Cannot use a constraint function on a sparse variable.")
E         RuntimeError: Cannot use a constraint function on a sparse variable.

../../../.pyenv/versions/3.6.9/envs/sg3/lib/python3.6/site-packages/tensorflow_core/python/keras/optimizer_v2/optimizer_v2.py:459: RuntimeError
```

</details>

Reply:
that's annoying that you cant constrain embeddings! another option is normalizing the embeddings in the `DistMultScore`

```
e1, r, e2 = inputs
e1 = tf.linalg.normalize(e1, axis=1,)
e2 = tf.linalg.normalize(e2, axis=1,)
```

Reply:
I'd personally prefer to not add complexity here (vs. waiting and doing the "right" thing of with `embeddings_constraint`), especially because AmpliGraph reports their best hyper parameters for DistMult are without normalisation: https://docs.ampligraph.org/en/1.2.0/experiments.html (`normalize_ent_emb: false` in each DistMult row).

Reply:
fair point; we can add it to the list of things different from the paper that will probably improve performance 😂

Reply:
Interesting, because in the paper they claim the opposite.

> The entity vectors are renormalized to have unit length...(it is an effective technique that empirically improved all the models..)

Reply:
Indeed; it's a peculiar mismatch: something to explore once TF supports constraining embeddings. (I also filed #980 to track this for us as well.)